File: compute-platform-spark/src/main/scala/org/lsw/compute/spark/api/data/DataSink.scala
package org.lsw.compute.spark.api.data

import org.apache.spark.sql.Dataset
import org.apache.spark.sql.streaming.StreamingQuery
import org.lsw.compute.spark.api.{Plugin, TransformBase}

/**
 * 输出的抽象
 *
 * @tparam IN
 */
abstract class DataSink[IN] extends Plugin[Unit] {

  protected var input: TransformBase[IN] = _

  def setInput(input: TransformBase[IN]): Unit = {
    this.input = input
  }

  /**
   * 一般输出实现
   */
  def save(): Either[StreamingQuery, Unit] = {

    if (this.input == null) {
      return Right()
    }

    if (config("autoCreate").toString.equals("true")) {
      createSinkIfNotExists()
    }
    val inData = this.input.transform()
    if (this.input.transform().isStreaming) {
      Left(streamSave(inData))
    } else {
      // We don't have to worry about batch output
      Right(batchSave(inData))
    }
  }

  /**
   * 批量dataframe的输出
   */
  private def batchSave(inData: Dataset[IN]): Unit = {

    config("saveMode").toString.toLowerCase match {
      case "overwrite" => BatchOverwrite(inData)
      case "append" => BatchAppend(inData)
      case other => throw new Exception(s"Unsupported types: $other")
    }
  }

  /**
   * 覆盖写
   */
  protected def BatchOverwrite(inData: Dataset[IN]): Unit

  /**
   * 增量写
   */
  protected def BatchAppend(inData: Dataset[IN]): Unit

  /**
   * 不存在则创建。
   */
  protected def createSinkIfNotExists(): Unit

  /**
   * 流式任务中的输出实现
   */
  protected def streamSave(inData: Dataset[IN]): StreamingQuery


}

File: compute-platform-core/src/main/java/org/lsw/core/operators/data/DataSink.java
package org.lsw.core.operators.data;

import org.lsw.core.model.ComputeData;

/**
 * 数据输出
 * @author liushengwei
 * @date 2021/09/23
 * @param <IN> 输入的数据类型
 */
public abstract class DataSink<IN> {
    /**
     * 需要输出的数据
     */
    public final ComputeData<IN> input;

    public DataSink(ComputeData<IN> input) {
        this.input = input;
    }

    /**
     * 将数据存储至外部
     * @param input 待存储的数据
     */
    public abstract void save(ComputeData<IN> input);
}
